### ISSUE TITLE
clarify command line binding rules

### ISSUE THREAD

**Issue by mr-c** [common-workflow-language/common-workflow-language#398]
http://www.commonwl.org/v1.0/CommandLineTool.html#Input_binding
4.1.3 say "If position is not specified, it is not added to the sorting key. "

But 5.1.2 http://www.commonwl.org/v1.0/CommandLineTool.html#CommandLineBinding says "position,	int, [not required], The sorting key. Default position is 0."

Here is what `cwltool` does: https://github.com/common-workflow-language/cwltool/blob/master/cwltool/builder.py#L59

**Comment by mr-c** [comment]
We should also mention the legality and usefulness of negative values for `position`

**Comment by tetron** [comment]
The idea was that if you specify position in a binding that's pushed down a level or two, for example a binding on a record field, it's not affected by a position applied to the record as a whole, unless explicitly listed.

If you have a record { a, b } and "a" is position 1 and "b" is position 2, the sort keys are [1, a] and [2, b].

If there's an implicit position 0 on the record the sort keys would be [0, 1, a] and [0, 2, b].  But I don't think that's what we want.

**Comment by mr-c** [comment]
I think we do want the implicit position 0 on the record as other inputBindings of the same "rank" may have an explicit position. 

Here are the principles as I understand them:

No priority due to the order written in the CWL document.
No priority between `arguments` and `inputs` level `inputBindings`

Default position is zero

First let us consider "flat" or simple items (no records/maps, no arrays)
If items share a position, then they are sorted by the elements to be added to the command line using the [Unicode Collation Algorithm](http://unicode.org/reports/tr10/)

For complex items (records, arrays) there are positions at the overall level (how an array or record compares with other items of the same rank or level) and at the element level (how the items of the array or record sort with respect to their peers). As always, the default `position` is zero. For these complex items there is a tie breaker with greater priority than the elements to be added to the command line: either the index of the item (in the case of arrays) or the name of the field (in the case of records).

**Comment by StarvingMarvin** [comment]
I agree with @mr-c here. Adding case where implicit 0 would behave differently that explicit 0 would be confusing to me.
